File: apps/web/lib/assistant_sections.ts
import { bodyText } from '@/lib/assistant_format';

export type AnswerSection = { id: string; title: string };

const SECTION_MD_RE = /^###\s+(.+)$/gm;
const SECTION_LEGACY_RE = /【([^】]+)】/g;

/** 小节锚点 id（与 Markdown h3 id 一致）。 */
export function sectionSlug(title: string): string {
  const base = title.trim().replace(/\s+/g, '-');
  const safe = base.replace(/[^\w\u4e00-\u9fff-]/g, '');
  return safe ? `sec-${safe}` : 'sec-section';
}

/** 从正文反解析 ### / 【】 小节列表。 */
export function parseAnswerSections(text: string): AnswerSection[] {
  const sections: AnswerSection[] = [];
  for (const m of text.matchAll(SECTION_MD_RE)) {
    const title = m[1]?.trim() ?? '';
    if (!title || title === '相关追问') break;
    sections.push({ id: sectionSlug(title), title });
  }
  if (sections.length) return sections;
  for (const m of text.matchAll(SECTION_LEGACY_RE)) {
    const title = m[1]?.trim() ?? '';
    if (!title || title === '相关追问') break;
    sections.push({ id: sectionSlug(title), title });
  }
  return sections;
}

/** done.sections 优先，否则从正文解析。 */
export function mergeAnswerSections(
  fromDone: AnswerSection[] | undefined,
  text: string,
): AnswerSection[] {
  if (fromDone?.length) {
    return fromDone.map((s) => ({
      id: s.id?.trim() || sectionSlug(s.title),
      title: s.title,
    }));
  }
  return parseAnswerSections(bodyText(text));
}

export const LEAD_SECTION_TITLES = new Set(['摘要', '本章概览', '卷概览']);

File: apps/mobile/lib/features/assistant/assistant_sections.dart
/// 小爱回答小节解析（对齐 Web `assistant_sections.ts`）。
library;

import 'assistant_format.dart';

class AnswerSection {
  const AnswerSection({required this.id, required this.title});

  final String id;
  final String title;

  factory AnswerSection.fromJson(Map<String, dynamic> j) => AnswerSection(
        id: (j['id'] ?? sectionSlug((j['title'] ?? '') as String)) as String,
        title: (j['title'] ?? '') as String,
      );
}

String sectionSlug(String title) {
  final base = title.trim().replaceAll(RegExp(r'\s+'), '-');
  final safe = base.replaceAll(RegExp(r'[^\w\u4e00-\u9fff-]'), '');
  return safe.isEmpty ? 'sec-section' : 'sec-$safe';
}

List<AnswerSection> parseAnswerSections(String text) {
  final sections = <AnswerSection>[];
  final md = RegExp(r'^###\s+(.+)$', multiLine: true);
  for (final m in md.allMatches(text)) {
    final title = m.group(1)?.trim() ?? '';
    if (title.isEmpty || title == '相关追问') break;
    sections.add(AnswerSection(id: sectionSlug(title), title: title));
  }
  if (sections.isNotEmpty) return sections;
  final legacy = RegExp(r'【([^】]+)】');
  for (final m in legacy.allMatches(text)) {
    final title = m.group(1)?.trim() ?? '';
    if (title.isEmpty || title == '相关追问') break;
    sections.add(AnswerSection(id: sectionSlug(title), title: title));
  }
  return sections;
}

List<AnswerSection> mergeAnswerSections(
  List<AnswerSection>? fromDone,
  String text,
) {
  if (fromDone != null && fromDone.isNotEmpty) {
    return fromDone
        .map(
          (s) => AnswerSection(
            id: s.id.trim().isEmpty ? sectionSlug(s.title) : s.id,
            title: s.title,
          ),
        )
        .toList();
  }
  return parseAnswerSections(bodyText(text));
}

const leadSectionTitles = {'摘要', '本章概览', '卷概览'};
